ikarus/finiteelements/mechanics/kirchhoffloveshell.hh

```diff
@@ -174,7 +163,7 @@ public:
    * \tparam RT The type representing the requested result.
    */
   template <template <typename, int, int> class RT>
-  requires(canProvideResultType<RT>())
+  requires(KirchhoffLoveShell::template supportsResultType<RT>())
```

Comment:
Do you really need `KirchhoffLoveShell::template ` in front? In the godbolt link it found the base class member function without problems

Reply:
I tried it like in the Godbolt example but it wouldn't compile (at least on GCC 12.0) I got a wall of compiler errors, this worked without a problem, I don't really know why

Reply:
Ok strange: here it works with gcc 12.2: https://godbolt.org/z/r38v14hef

Reply:
```
[build] /workspaces/ikarus/tests/src/testlinearelasticity.cpp:36:26:   required from here
[build] /workspaces/ikarus/tests/src/testfeelement.hh:98:19: error: static assertion failed
[build]    98 |     static_assert(requires {
[build]       |                   ^~~~~~~~~~
[build]    99 |       fes[0].template calculateAt<RT::template Rebind>(requirements, Dune::FieldVector<double, gridDim>());
[build]       |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build]   100 |     });
[build]       |     ~              
[build] /workspaces/ikarus/tests/src/testfeelement.hh:98:19: note: ‘false’ evaluates to false
[build] /workspaces/ikarus/tests/src/testfeelement.hh: In instantiation of ‘testFEElement<main(int, char**)::<lambda(const Ikarus::YoungsModulusAndPoissonsRatio&)>&, Dune::Functions::BasisFactory::power<2, lagrange<2>()::<lambda(const auto:86&)>, FlatInterleaved>(lagrange<2>()::<lambda(const auto:86&)>&&, const FlatInterleaved&)::<lambda(const auto:471&)>, Dune::Geo::ReferenceElement<Dune::Geo::ReferenceElementImplementation<double, 2> >, Ikarus::Skills<>, Ikarus::ScalarAffordance, Ikarus::VectorAffordance, Ikarus::MatrixAffordance, <lambda(auto:477&, auto:478&, auto:479&, auto:480&)>&, <lambda(auto:481&, auto:482&, auto:483&, auto:484&)>&, <lambda(auto:485&, auto:486&, auto:487&, auto:488&)>&, <lambda(auto:494&, auto:495&, auto:496&, auto:497&)>&>(const Dune::Functions::BasisFactory::power<2, lagrange<2>()::<lambda(const auto:86&)>, FlatInterleaved>(lagrange<2>()::<lambda(const auto:86&)>&&, const FlatInterleaved&)::<lambda(const auto:471&)>&, const std::string&, const CornerDistortionFlag&, const Dune::Geo::ReferenceElement<Dune::Geo::ReferenceElementImplementation<double, 2> >&, main(int, char**)::<lambda(const Ikarus::YoungsModulusAndPoissonsRatio&)>&, Ikarus::Skills<>&&, Ikarus::AffordanceCollection<Ikarus::ScalarAffordance, Ikarus::VectorAffordance, Ikarus::MatrixAffordance>, <lambda(auto:477&, auto:478&, auto:479&, auto:480&)>&, <lambda(auto:481&, auto:482&, auto:483&, auto:484&)>&, <lambda(auto:485&, auto:486&, auto:487&, auto:488&)>&, <lambda(auto:494&, auto:495&, auto:496&, auto:497&)>&)::<lambda(RT)> [with RT = Ikarus::ResultTypes::linearStress<double, 1, 1>]’:
[build] /dune/dune-common/dune/common/hybridutilities.hh:207:36:   required from ‘constexpr void Dune::Hybrid::Impl::forEachIndex(Range&&, F&&, std::integer_sequence<Index, i ...>) [with Range = std::tuple<Ikarus::ResultTypes::linearStress<double, 1, 1> >; F = testFEElement<main(int, char**)::<lambda(const Ikarus::YoungsModulusAndPoissonsRatio&)>&, Dune::Functions::BasisFactory::power<2, lagrange<2>()::<lambda(const auto:86&)>, FlatInterleaved>(lagrange<2>()::<lambda(const auto:86&)>&&, const FlatInterleaved&)::<lambda(const auto:471&)>, Dune::Geo::ReferenceElement<Dune::Geo::ReferenceElementImplementation<double, 2> >, Ikarus::Skills<>, Ikarus::ScalarAffordance, Ikarus::VectorAffordance, Ikarus::MatrixAffordance, <lambda(auto:477&, auto:478&, auto:479&, auto:480&)>&, <lambda(auto:481&, auto:482&, auto:483&, auto:484&)>&, <lambda(auto:485&, auto:486&, auto:487&, auto:488&)>&, <lambda(auto:494&, auto:495&, auto:496&, auto:497&)>&>(const Dune::Functions::BasisFactory::power<2, lagrange<2>()::<lambda(const auto:86&)>, FlatInterleaved>(lagrange<2>()::<lambda(const auto:86&)>&&, const FlatInterleaved&)::<lambda(const auto:471&)>&, const std::string&, const CornerDistortionFlag&, const Dune::Geo::ReferenceElement<Dune::Geo::ReferenceElementImplementation<double, 2> >&, main(int, char**)::<lambda(const Ikarus::YoungsModulusAndPoissonsRatio&)>&, Ikarus::Skills<>&&, Ikarus::AffordanceCollection<Ikarus::ScalarAffordance, Ikarus::VectorAffordance, Ikarus::MatrixAffordance>, <lambda(auto:477&, auto:478&, auto:479&, auto:480&)>&, <lambda(auto:481&, auto:482&, auto:483&, auto:484&)>&, <lambda(auto:485&, auto:486&, auto:487&, auto:488&)>&, <lambda(auto:494&, auto:495&, auto:496&, auto:497&)>&)::<lambda(RT)>; Index = long unsigned int; Index ...i = {0}]’
```

Reply:
I think you have created a circular header inclusion and in some path, when you include 
`ikarus/finiteelements/mechanics/linearelastic.hh`

it does not now the content of `#include <ikarus/finiteelements/feresulttypes.hh>`.
Can you try to put the result type base class in a seperate file and try that? I compared to godbolt but I can not find a difference otherwise.

If you add the `Class::template foo<RT>` you delay this lookup and then it worked for your case I suppose.

Reply:
Sorry, I couldn't get it to work even when putting in a different file. 
I think the compiler dosn't know about the base class implementation when parsing the requries
Clang++18 says here: 
```
 In file included from /workspaces/ikarus/tests/src/testlinearelasticity.cpp:7:
[build] /workspaces/ikarus/tests/src/testfeelement.hh:98:19: error: static assertion failed due to requirement 'requires { <<error-expression>>; }'
[build]    98 |     static_assert(requires {
[build]       |                   ^~~~~~~~~~
[build]    99 |       fes[0].template calculateAt<RT::template Rebind>(requirements, Dune::FieldVector<double, gridDim>());
[build]       |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build]   100 |     });
[build]       |     ~
```
So i think we have to live with my workaorund, or do you have another idea?

Reply:
This static assert will never work. But

`there are no arguments to ‘supportsResultType’ that depend on a template parameter, so a declaration of ‘supportsResultType’ must be available [-fpermissive]` should be gone

Reply:
Ah no wait this is for `linearelasticity` but not for kirchhoff love. So this should a working overload, shouldn't it?

Reply:
It does not work with both compilers.
I think we have to understand the problem, even with your workaround the problem may reappear somewhere else since somehow the visibility of `supportsResultType` is strange. What happens if you write `    static_assert(d.supportsResultType<RT1>());` in testlinearelasticity?

Reply:
Maybe the problem is that i have parametriced the baseclass with the element itself, therefore the type should be incomplete at the time where the compiler sees the methods the first time, i will check if we have to include `Derived` in the BaseClass